File: Gold/modules/vumanchu.py
"""
VuManChu Cipher B — three-component confirmation engine.

Components:
  WT1/WT2     WaveTrend momentum oscillator (matches dashboard divergence.js)
  Money Flow  Volume-weighted directional pressure, scaled –100 to +100
  VWAP slope  Session VWAP momentum exhaustion at the zone

VWAP is NOT used as a directional bias (above/below price = bullish/bearish).
That is the wrong read. The VWAP right-angle PRICE LEVELS are handled by the
session engine as historical confluence zones. Here, VWAP slope tells us
whether the momentum pushing price INTO the zone is running out of energy:

  For a LONG zone (looking to buy at support):
    Good: VWAP slope was falling but is now flattening or turning → bearish
          momentum exhausting as price reaches the zone (sellers tiring)
    Good: VWAP slope actively turning positive → reversal underway

  For a SHORT zone (looking to sell at resistance):
    Good: VWAP slope was rising but is now flattening or turning → bullish
          momentum exhausting as price reaches the zone (buyers tiring)
    Good: VWAP slope actively turning negative → reversal underway

Gold-specific money flow behaviour:
  A money-flow SPIKE as price hits a zone signals EXHAUSTION, not continuation.
  (opposite to Bitcoin — sellers/buyers forced in, then snap-back)

WT signal priority (most specific first):
  1. OVERSOLD/OVERBOUGHT — WT1 beyond ±60 at the zone: direct exhaustion read.
     More reliable for gold than a simple crossover which flips repeatedly.
  2. ZONE-ENTRY DIVERGENCE — price makes a new extreme inside the zone but WT
     does not confirm it (classic Cipher B setup). Only counted when entry_time
     is provided so divergence is measured from the first GP touch onwards.
  3. WT1 > WT2 crossover — weakest signal, only used when neither above applies.

Entry fires only when components_aligned >= min_required (default 2).
Direction alignment is checked against the zone's expected trade direction.
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Optional


@dataclass
class VuManChuSignal:
    direction: str           # LONG | SHORT | NEUTRAL
    confidence: str          # HIGH | MEDIUM | LOW
    wt1: float
    wt2: float
    wt_signal: str           # OVERSOLD | OVERBOUGHT | DIVERGENCE_BULL | DIVERGENCE_BEAR | BULLISH | BEARISH | NEUTRAL
    mf_value: float
    mf_signal: str           # BULLISH_EXHAUSTION | BEARISH_EXHAUSTION | BULLISH | BEARISH | NEUTRAL
    vwap_signal: str         # EXHAUSTION | REVERSAL | NEUTRAL
    components_aligned: int  # 0–3
    reason: str
    zone_entry_bar_idx: int = 0   # bar index in the passed series when price entered the GP


# ── EMA / SMA helpers ────────────────────────────────────────────────────────

def _ema(vals: list[float], period: int) -> list[float]:
    if not vals or period <= 0:
        return []
    k, res = 2.0 / (period + 1), [vals[0]]
    for v in vals[1:]:
        res.append(v * k + res[-1] * (1 - k))
    return res


def _sma(vals: list[float], period: int) -> list[float]:
    out = []
    for i in range(len(vals)):
        if i < period - 1:
            out.append(float('nan'))
        else:
            out.append(sum(vals[i - period + 1: i + 1]) / period)
    return out


# ── WT series ────────────────────────────────────────────────────────────────

def _wt_series(bars: list[dict], n1: int = 10, n2: int = 21
               ) -> tuple[list[float], list[float]]:
    closes = [float(b['close']) for b in bars]
    highs  = [float(b['high'])  for b in bars]
    lows   = [float(b['low'])   for b in bars]
    hlc3   = [(h + l + c) / 3 for h, l, c in zip(highs, lows, closes)]

    esa = _ema(hlc3, n1)
    d   = _ema([abs(h - e) for h, e in zip(hlc3, esa)], n1)
    ci  = [(h - e) / (0.015 * dv) if dv else 0.0
           for h, e, dv in zip(hlc3, esa, d)]

    wt1 = _ema(ci, n2)
    wt2 = _sma(wt1, 4)
    return wt1, wt2


# ── Money flow ───────────────────────────────────────────────────────────────

def _money_flow(bars: list[dict], period: int = 14) -> list[float]:
    """
    Directional volume pressure per bar: (close-open)/(high-low) × volume.
    Smoothed with EMA and normalised to [–100, +100].
    """
    raw = []
    for b in bars:
        rng = b['high'] - b['low'] + 0.001
        raw.append((b['close'] - b['open']) / rng * b.get('tick_volume', 1))

    peak = max(abs(v) for v in raw) if raw else 1.0
    if peak > 0:
        raw = [v / peak * 100 for v in raw]
    return _ema(raw, period)


# ── VWAP slope exhaustion ────────────────────────────────────────────────────

def _vwap_exhaustion(bars: list[dict], zone_direction: str,
                     window: int = 20) -> str:
    """
    Reads the session VWAP slope to determine whether the momentum carrying
    price INTO the zone is running out of energy.

    Rolling VWAP is computed cumulatively from bar[0]. We measure slope
    change: early-window slope vs late-window slope. If the slope was strong
    in the move's direction and is now flattening or reversing, the momentum
    is exhausting — which is exactly the confirmation needed for a reversal
    entry at the zone.

    Returns:
      EXHAUSTION  — slope was strong, now flattening → fuel running out
      REVERSAL    — slope has actively turned the other way → confirmation
      NEUTRAL     — no clear exhaustion signal
    """
    if len(bars) < window + 5:
        return 'NEUTRAL'

    # Build cumulative VWAP series
    cum_tpv = cum_vol = 0.0
    vwap_series: list[float] = []
    for b in bars:
        tp = (b['high'] + b['low'] + b['close']) / 3
        v  = b.get('tick_volume', 1)
        cum_tpv += tp * v; cum_vol += v
        vwap_series.append(cum_tpv / cum_vol if cum_vol else tp)

    recent = vwap_series[-window:]
    half   = window // 2

    early_slope = recent[half] - recent[0]       # first half of window
    late_slope  = recent[-1]  - recent[half]     # second half of window

    if zone_direction == 'long':
        # Price falling into support zone — we want bearish VWAP slope to be exhausting
        if early_slope < 0 and late_slope > 0:
            return 'REVERSAL'      # VWAP slope actively turned up
        if early_slope < 0 and abs(late_slope) < abs(early_slope) * 0.45:
            return 'EXHAUSTION'    # downward momentum fading fast
    else:
        # Price rising into resistance zone — we want bullish VWAP slope to be exhausting
        if early_slope > 0 and late_slope < 0:
            return 'REVERSAL'      # VWAP slope actively turned down
        if early_slope > 0 and abs(late_slope) < abs(early_slope) * 0.45:
            return 'EXHAUSTION'    # upward momentum fading fast

    return 'NEUTRAL'


# ── Divergence ───────────────────────────────────────────────────────────────

def _divergence(closes: list[float], wt: list[float], n: int = 5,
                start_idx: int = 0) -> str:
    """
    Detect price/WT divergence.

    start_idx: first bar to include in the divergence window (defaults to 0 =
    whole series). When zone_entry_bar_idx is provided, only bars from zone
    entry onwards are used so we detect divergence *at the zone* rather than
    anywhere in the lookback window.

    At least n+2 bars must be available after start_idx; if not, falls back to
    checking the last n+2 bars of the full series so we always have a reading.
    """
    c = closes[start_idx:] if start_idx > 0 else closes
    w = wt[start_idx:]     if start_idx > 0 else wt

    # Fallback to tail of full series if entry window is too short
    if len(c) < n + 2:
        c = closes[-(n + 2):]
        w = wt[-(n + 2):]

    if len(c) < n + 2 or len(w) < n + 2:
        return 'NONE'

    wt_peak    = max(w[-n:])
    wt_trou    = min(w[-n:])
    price_peak = max(c[-n:])
    price_trou = min(c[-n:])

    # Bearish div: price at/near peak but WT rolling lower
    if c[-1] >= price_peak * 0.999 and w[-1] < wt_peak * 0.90:
        return 'DIVERGENCE_BEAR'
    # Bullish div: price at/near trough but WT turning higher
    if c[-1] <= price_trou * 1.001 and w[-1] > wt_trou * 0.90:
        return 'DIVERGENCE_BULL'
    return 'NONE'


# WT levels for gold — beyond ±60 indicates genuine exhaustion at a zone.
# A crossover alone is too noisy on a 5m chart; oversold/overbought is a
# much cleaner read of momentum exhaustion as price sits in the GP.
WT_OVERSOLD  = -60.0
WT_OVERBOUGHT = 60.0


# ── Public API ────────────────────────────────────────────────────────────────

def compute_vumanchu(bars: list[dict], zone_direction: str,
                     n1: int = 10, n2: int = 21, mf_period: int = 14,
                     min_components: int = 2,
                     entry_time: Optional[float] = None) -> VuManChuSignal:
    """
    bars:           5m or 15m bars, at least 40 recommended (chronological).
    zone_direction: 'long' or 'short' — expected trade direction.
    entry_time:     Unix timestamp (float) of when price first entered the GP
                    window. When supplied, divergence is measured from that bar
                    onwards so we catch Cipher B setups that form *at the zone*
                    rather than anywhere in the lookback window.
    """
    if len(bars) < n2 + n1:
        return VuManChuSignal(
            direction='NEUTRAL', confidence='LOW',
            wt1=0.0, wt2=0.0, wt_signal='NEUTRAL',
            mf_value=0.0, mf_signal='NEUTRAL',
            vwap_signal='NEUTRAL', components_aligned=0,
            reason='Insufficient bars',
        )

    closes       = [float(b['close']) for b in bars]
    wt1_s, wt2_s = _wt_series(bars, n1, n2)
    mf_s         = _money_flow(bars, mf_period)
    vwap_sig     = _vwap_exhaustion(bars, zone_direction)

    wt1   = wt1_s[-1] if wt1_s else 0.0
    wt2_v = wt2_s[-1] if wt2_s and not (isinstance(wt2_s[-1], float) and
                                          wt2_s[-1] != wt2_s[-1]) else 0.0
    mf    = mf_s[-1]  if mf_s  else 0.0

    # ── Resolve zone-entry bar index ──────────────────────────────────────────
    # Find the first bar whose timestamp is >= entry_time so divergence is
    # computed only over bars captured while price was inside the GP window.
    zone_entry_bar_idx = 0
    if entry_time is not None:
        for i, b in enumerate(bars):
            if b.get('time', 0) >= entry_time:
                zone_entry_bar_idx = i
                break

    # ── WT signal — priority order ────────────────────────────────────────────
    # 1. OVERSOLD / OVERBOUGHT — most direct exhaustion read for gold.
    #    WT1 < -60 at a long zone: sellers fully extended, reversal likely.
    #    WT1 > +60 at a short zone: buyers fully extended, reversal likely.
    # 2. Zone-entry divergence (Cipher B): price makes new extreme inside the GP
    #    but WT does not confirm — measured from zone_entry_bar_idx onwards.
    # 3. WT1/WT2 crossover — weakest; only used when neither above fires.
    if zone_direction == 'long' and wt1 <= WT_OVERSOLD:
        wt_sig = 'OVERSOLD'
    elif zone_direction == 'short' and wt1 >= WT_OVERBOUGHT:
        wt_sig = 'OVERBOUGHT'
    else:
        div = _divergence(closes, wt1_s, n=5, start_idx=zone_entry_bar_idx)
        if div != 'NONE':
            wt_sig = div
        elif wt1 > wt2_v:
            wt_sig = 'BULLISH'
        elif wt1 < wt2_v:
            wt_sig = 'BEARISH'
        else:
            wt_sig = 'NEUTRAL'

    # ── Money flow (gold: spike = exhaustion) ─────────────────────────────────
    prev5  = mf_s[-6:-1] if len(mf_s) >= 6 else mf_s[:-1]
    mf_max = max(prev5) if prev5 else mf
    mf_min = min(prev5) if prev5 else mf

    if mf_max > 30 and mf < mf_max * 0.7:
        mf_sig = 'BEARISH_EXHAUSTION'   # bullish spike rolled over → sellers tapped out at resistance
    elif mf_min < -30 and mf > mf_min * 0.7:
        mf_sig = 'BULLISH_EXHAUSTION'   # bearish spike rolling back → buyers tapped out at support
    elif mf > 20:
        mf_sig = 'BULLISH'
    elif mf < -20:
        mf_sig = 'BEARISH'
    else:
        mf_sig = 'NEUTRAL'

    # ── Count aligned components ──────────────────────────────────────────────
    aligned = 0
    notes: list[str] = []

    vwap_confirmed = vwap_sig in ('EXHAUSTION', 'REVERSAL')

    if zone_direction == 'long':
        if wt_sig in ('OVERSOLD', 'BULLISH', 'DIVERGENCE_BULL'):
            aligned += 1; notes.append(f'WT {wt_sig}')
        if mf_sig in ('BULLISH_EXHAUSTION', 'BULLISH'):
            aligned += 1; notes.append(f'MF {mf_sig}')
        if vwap_confirmed:
            aligned += 1; notes.append(f'VWAP slope {vwap_sig}')
    else:
        if wt_sig in ('OVERBOUGHT', 'BEARISH', 'DIVERGENCE_BEAR'):
            aligned += 1; notes.append(f'WT {wt_sig}')
        if mf_sig in ('BEARISH_EXHAUSTION', 'BEARISH'):
            aligned += 1; notes.append(f'MF {mf_sig}')
        if vwap_confirmed:
            aligned += 1; notes.append(f'VWAP slope {vwap_sig}')

    if aligned >= 3:
        confidence = 'HIGH';   direction = zone_direction.upper()
    elif aligned >= min_components:
        confidence = 'MEDIUM'; direction = zone_direction.upper()
    else:
        confidence = 'LOW';    direction = 'NEUTRAL'

    return VuManChuSignal(
        direction=direction, confidence=confidence,
        wt1=round(wt1, 2), wt2=round(wt2_v, 2), wt_signal=wt_sig,
        mf_value=round(mf, 1), mf_signal=mf_sig,
        vwap_signal=vwap_sig,
        components_aligned=aligned,
        reason=' · '.join(notes) or 'No alignment',
        zone_entry_bar_idx=zone_entry_bar_idx,
    )

File: js/vumanchu.js
/**
 * vumanchu.js — VuManChu Cipher B: central shared utility
 *
 * Pure functions only — no imports from dashboard state, no side effects.
 * Designed to work in any context: server monitoring loop, browser signal
 * engine, Python-via-subprocess, or future bots.
 *
 * Callers pass a bars array and choose which components they need.
 * Timeframe is caller-controlled — pass M1, M5, M15, or resampled bars.
 * Use resampleBars(m1Bars, 5) to get M5-quality signals from M1 data.
 *
 * Components:
 *   WT    WaveTrend oscillator (WT1 line + WT2 signal line)
 *   MF    Money Flow — directional volume pressure scaled –100 to +100
 *   VWAP  Dual-mode: slope exhaustion + price-vs-VWAP oscillator divergence
 *
 * Divergence types (applied to WT and VWAP oscillators):
 *   DIVERGENCE_BULL  Price lower low, oscillator higher low → reversal long
 *   DIVERGENCE_BEAR  Price higher high, oscillator lower high → reversal short
 *   HIDDEN_BULL      Price higher low, oscillator lower low → trend continuation long
 *   HIDDEN_BEAR      Price lower high, oscillator higher high → trend continuation short
 *
 * Entry verdict from assessEntry():
 *   'agree'   — 2+ components confirm the trade direction
 *   'neutral' — oscillators not committed either way
 *   'oppose'  — oscillators actively contradict the trade direction
 *
 * Note on Money Flow for FX:
 *   FX tick_volume (price updates per bar) is not true traded volume.
 *   MF is included and useful as a directional pressure proxy, but weight
 *   it lower than for Gold where futures tick volume is a reliable proxy.
 *   Set opts.useMF = false to skip it entirely for a given call.
 */

// ── Shared compute (single source of truth) ──────────────────────────────────
// EMA/SMA, WaveTrend, Money Flow and VWAP live in the vumanchuCore brick so this
// module and asiaRangeEngine share ONE implementation (the divide-by-zero guard
// was the only thing that had drifted; core standardizes it on 1e-10, which also
// suppresses flat-market float-noise spikes). Re-exported under the original
// names so existing importers (signal.js computeWT, server.js, gold backtest)
// don't change.
import {
  ema, sma, computeWaveTrend, computeMoneyFlow, computeVWAP as _computeVWAP,
} from './vumanchuCore.js';

export { ema, sma };

// ── Bar resampling ────────────────────────────────────────────────────────────

/**
 * Resample an array of M1 bars into N-minute bars.
 * Bars must be in chronological order (oldest first).
 * Incomplete final group is discarded to avoid a forming-bar bias.
 *
 * Each bar: { open, high, low, close, volume? }
 * Returns same shape, fewer bars.
 */
export function resampleBars(bars, n) {
  if (!bars?.length || n <= 1) return bars ?? [];
  const out = [];
  for (let i = 0; i + n <= bars.length; i += n) {
    const group = bars.slice(i, i + n);
    out.push({
      open:   group[0].open,
      high:   Math.max(...group.map(b => b.high)),
      low:    Math.min(...group.map(b => b.low)),
      close:  group[group.length - 1].close,
      volume: group.reduce((s, b) => s + (b.volume ?? b.tick_volume ?? 1), 0),
    });
  }
  return out;
}

// ── WaveTrend oscillator ──────────────────────────────────────────────────────
// Classic VuManChu Cipher B formula. Parameters:
//   n1=10  channel length (ESA + deviation smoothing period)
//   n2=21  WT1 smoothing period
//   sp=4   WT2 SMA signal line period

export const computeWT = (bars, opts = {}) => computeWaveTrend(bars, opts);

// ── Money Flow ────────────────────────────────────────────────────────────────
// Directional volume pressure: (close − open) / range × volume, EMA-smoothed.
// Positive = buying pressure, negative = selling pressure.
// Scaled to ±100 against the peak value in the series.
//
// FX note: use tick_volume (price updates per bar) as the volume proxy.

export const computeMF = (bars, opts = {}) => computeMoneyFlow(bars, opts);

// ── VWAP ──────────────────────────────────────────────────────────────────────

/**
 * Session VWAP (cumulative from bar 0). Returns { vwap, osc }. Delegates to the
 * vumanchuCore brick (single source of truth).
 */
export const computeVWAP = (bars) => _computeVWAP(bars);

/**
 * VWAP slope exhaustion — detects whether the momentum carrying price into a
 * zone is fading. Compares early-window slope vs late-window slope.
 *
 * Returns 'EXHAUSTION' | 'REVERSAL' | 'NEUTRAL'
 */
export function vwapExhaustion(bars, direction, { window = 20 } = {}) {
  if (bars.length < window + 5) return 'NEUTRAL';
  const { vwap } = computeVWAP(bars);
  const recent = vwap.slice(-window);
  const half   = Math.floor(window / 2);
  const earlySlope = recent[half]    - recent[0];
  const lateSlope  = recent[window - 1] - recent[half];

  if (direction === 'long') {
    if (earlySlope < 0 && lateSlope > 0)                               return 'REVERSAL';
    if (earlySlope < 0 && Math.abs(lateSlope) < Math.abs(earlySlope) * 0.45) return 'EXHAUSTION';
  } else {
    if (earlySlope > 0 && lateSlope < 0)                               return 'REVERSAL';
    if (earlySlope > 0 && Math.abs(lateSlope) < Math.abs(earlySlope) * 0.45) return 'EXHAUSTION';
  }
  return 'NEUTRAL';
}

// ── Swing point detection ─────────────────────────────────────────────────────

/**
 * Find local swing highs and lows in a series.
 * Returns { highs: [{i, v}], lows: [{i, v}] } in chronological order.
 * left/right: number of strictly-lower/higher bars required on each side.
 */
export function findSwings(series, { left = 3, right = 2 } = {}) {
  const highs = [], lows = [];
  for (let i = left; i < series.length - right; i++) {
    const v = series[i];
    let isH = true, isL = true;
    for (let j = 1; j <= left;  j++) { if (series[i - j] >= v) { isH = false; } if (series[i - j] <= v) { isL = false; } }
    for (let j = 1; j <= right; j++) { if (series[i + j] >= v) { isH = false; } if (series[i + j] <= v) { isL = false; } }
    if (isH) highs.push({ i, v });
    if (isL) lows.push({ i, v });
  }
  return { highs, lows };
}

// ── Structural divergence ─────────────────────────────────────────────────────

const OSC_MIN_DIFF = 2.0; // min oscillator unit gap between swings to count

/**
 * Detect structural divergence between price and an oscillator.
 *
 * startIdx: bar index from which to begin (use 0 for full series, or the
 *   index where price first entered a zone for zone-specific divergences).
 *
 * Returns 'DIVERGENCE_BULL' | 'DIVERGENCE_BEAR' |
 *         'HIDDEN_BULL'     | 'HIDDEN_BEAR'     | 'NONE'
 */
export function detectDivergence(closes, oscillator, {
  startIdx  = 0,
  oscWindow = 2,
  minGap    = 5,
  minDiff   = OSC_MIN_DIFF,
  swingLeft = 3,
  swingRight = 2,
} = {}) {
  let c = startIdx > 0 ? closes.slice(startIdx)     : closes;
  let o = startIdx > 0 ? oscillator.slice(startIdx) : oscillator;

  const minBars = swingLeft + swingRight + minGap + 2;
  if (c.length < minBars || o.length < minBars) {
    // Zone-entry window too short — fall back to full series tail
    c = closes.slice(-minBars);
    o = oscillator.slice(-minBars);
    if (c.length < minBars) return 'NONE';
  }

  const n = Math.min(c.length, o.length);
  c = c.slice(0, n); o = o.slice(0, n);

  const { highs: pH, lows: pL } = findSwings(c, { left: swingLeft, right: swingRight });

  // Read oscillator extreme in a ±oscWindow bar window around a price swing
  const oscNear = (idx, takeMax) => {
    const seg = o.slice(Math.max(0, idx - oscWindow), Math.min(n, idx + oscWindow + 1));
    if (!seg.length) return null;
    return takeMax ? Math.max(...seg) : Math.min(...seg);
  };

  // Divergence at swing HIGHS
  if (pH.length >= 2) {
    const { i: i1, v: ph1 } = pH[pH.length - 2];
    const { i: i2, v: ph2 } = pH[pH.length - 1];
    if (i2 - i1 >= minGap) {
      const oh1 = oscNear(i1, true), oh2 = oscNear(i2, true);
      if (oh1 != null && oh2 != null) {
        if (ph2 > ph1 && oh1 - oh2 >= minDiff) return 'DIVERGENCE_BEAR';
        if (ph2 < ph1 && oh2 - oh1 >= minDiff) return 'HIDDEN_BEAR';
      }
    }
  }

  // Divergence at swing LOWS
  if (pL.length >= 2) {
    const { i: i1, v: pl1 } = pL[pL.length - 2];
    const { i: i2, v: pl2 } = pL[pL.length - 1];
    if (i2 - i1 >= minGap) {
      const ol1 = oscNear(i1, false), ol2 = oscNear(i2, false);
      if (ol1 != null && ol2 != null) {
        if (pl2 < pl1 && ol2 - ol1 >= minDiff) return 'DIVERGENCE_BULL';
        if (pl2 > pl1 && ol1 - ol2 >= minDiff) return 'HIDDEN_BULL';
      }
    }
  }

  return 'NONE';
}

// ── Entry quality assessment ──────────────────────────────────────────────────

/**
 * Full VuManChu entry assessment for a given set of bars and trade direction.
 *
 * opts:
 *   obLevel      {number}  WT1 overbought threshold (default 60)
 *   osLevel      {number}  WT1 oversold threshold (default -60)
 *   minComponents{number}  components needed for 'agree' (default 2)
 *   useMF        {boolean} include Money Flow component (default true)
 *   useVWAP      {boolean} include VWAP component (default true)
 *   startIdx     {number}  bar from which to measure divergence (0 = full series)
 *   n1/n2/sp     {number}  WT parameters
 *
 * Returns:
 *   signal      'agree' | 'neutral' | 'oppose'
 *   confidence  'HIGH' | 'MEDIUM' | 'LOW'
 *   components  number of aligned components (0–3)
 *   wt          { value, signal }
 *   mf          { value, signal }   (null when useMF=false)
 *   vwap        { signal, divergence } (null when useVWAP=false)
 *   reason      human-readable summary string
 *   warnings    string[]
 */
export function assessEntry(bars, direction, opts = {}) {
  const {
    obLevel       = 60,
    osLevel       = -60,
    minComponents = 2,
    useMF         = true,
    useVWAP       = true,
    startIdx      = 0,
    n1 = 10, n2 = 21, sp = 4,
    mfPeriod = 14,
    vwapWindow = 20,
  } = opts;

  const isLong = direction === 'long';
  const neutral = (reason, warnings = []) => ({
    signal: 'neutral', confidence: 'LOW', components: 0,
    wt: null, mf: null, vwap: null, reason, warnings,
  });

  if (!bars?.length || bars.length < n1 + n2) {
    return neutral('Insufficient bars');
  }

  const closes = bars.map(b => parseFloat(b.close));

  // ── WaveTrend ──────────────────────────────────────────────────────────────
  const { wt1: wt1s, wt2: wt2s } = computeWT(bars, { n1, n2, sp });
  const wt1Val = wt1s[wt1s.length - 1] ?? 0;
  const wt2Val = wt2s[wt2s.length - 1];
  const wt2Num = isNaN(wt2Val) ? 0 : wt2Val;

  let wtSig;
  if      (isLong  && wt1Val <= osLevel) wtSig = 'OVERSOLD';
  else if (!isLong && wt1Val >= obLevel) wtSig = 'OVERBOUGHT';
  else {
    const wtDiv = detectDivergence(closes, wt1s, { startIdx });
    if   (wtDiv !== 'NONE') wtSig = wtDiv;
    else if (wt1Val > wt2Num)  wtSig = 'BULLISH';
    else if (wt1Val < wt2Num)  wtSig = 'BEARISH';
    else                       wtSig = 'NEUTRAL';
  }

  // ── Money Flow ─────────────────────────────────────────────────────────────
  let mfSig = null, mfVal = 0;
  if (useMF) {
    const mfS  = computeMF(bars, { period: mfPeriod });
    mfVal      = mfS[mfS.length - 1] ?? 0;
    const prev5 = mfS.slice(-6, -1);
    const mfMax = prev5.length ? Math.max(...prev5) : mfVal;
    const mfMin = prev5.length ? Math.min(...prev5) : mfVal;

    if      (mfMax >  30 && mfVal < mfMax * 0.7)  mfSig = 'BEARISH_EXHAUSTION';
    else if (mfMin < -30 && mfVal > mfMin * 0.7)  mfSig = 'BULLISH_EXHAUSTION';
    else if (mfVal >  20)                          mfSig = 'BULLISH';
    else if (mfVal < -20)                          mfSig = 'BEARISH';
    else                                           mfSig = 'NEUTRAL';
  }

  // ── VWAP ───────────────────────────────────────────────────────────────────
  let vwapSig = null, vwapDiv = 'NONE';
  if (useVWAP) {
    const { osc } = computeVWAP(bars);
    vwapSig = vwapExhaustion(bars, direction, { window: vwapWindow });
    vwapDiv = detectDivergence(closes, osc, { startIdx });
  }

  // ── Count aligned components ───────────────────────────────────────────────
  const LONG_WT_OK   = ['OVERSOLD', 'BULLISH', 'DIVERGENCE_BULL', 'HIDDEN_BULL'];
  const SHORT_WT_OK  = ['OVERBOUGHT', 'BEARISH', 'DIVERGENCE_BEAR', 'HIDDEN_BEAR'];
  const LONG_MF_OK   = ['BULLISH_EXHAUSTION', 'BULLISH'];
  const SHORT_MF_OK  = ['BEARISH_EXHAUSTION', 'BEARISH'];
  const LONG_MF_BAD  = ['BEARISH_EXHAUSTION', 'BEARISH'];
  const SHORT_MF_BAD = ['BULLISH_EXHAUSTION', 'BULLISH'];

  const notes    = [];
  const warnings = [];
  let   aligned  = 0;
  let   opposing = 0;

  // WT
  const wtOk  = isLong  ? LONG_WT_OK.includes(wtSig)  : SHORT_WT_OK.includes(wtSig);
  const wtBad = isLong  ? SHORT_WT_OK.includes(wtSig) : LONG_WT_OK.includes(wtSig);
  if (wtOk)  { aligned++;  notes.push(`WT ${wtSig}`); }
  if (wtBad) { opposing++; warnings.push(`WT ${wtSig} opposing`); }

  // MF
  if (useMF && mfSig) {
    const mfOk  = isLong  ? LONG_MF_OK.includes(mfSig)  : SHORT_MF_OK.includes(mfSig);
    const mfBad = isLong  ? LONG_MF_BAD.includes(mfSig) : SHORT_MF_BAD.includes(mfSig);
    if (mfOk)  { aligned++;  notes.push(`MF ${mfSig}`); }
    if (mfBad) { opposing++; warnings.push(`MF ${mfSig} opposing`); }
  }

  // VWAP
  if (useVWAP && vwapSig) {
    const vwapMomentumOk = vwapSig !== 'NEUTRAL';  // EXHAUSTION or REVERSAL — valid for both directions
    const vwapDivLong    = ['DIVERGENCE_BULL', 'HIDDEN_BULL'].includes(vwapDiv);
    const vwapDivShort   = ['DIVERGENCE_BEAR', 'HIDDEN_BEAR'].includes(vwapDiv);
    const vwapConfirmed  = isLong ? (vwapMomentumOk || vwapDivLong)  : (vwapMomentumOk || vwapDivShort);
    const vwapOpposed    = isLong ? vwapDivShort : vwapDivLong;
    const vwapLabel      = vwapDiv !== 'NONE' ? `VWAP ${vwapDiv}` : `VWAP ${vwapSig}`;
    if (vwapConfirmed && !vwapOpposed) { aligned++;  notes.push(vwapLabel); }
    if (vwapOpposed)                   { opposing++; warnings.push(`${vwapLabel} opposing`); }
  }

  // ── Verdict ────────────────────────────────────────────────────────────────
  const maxComponents = 1 + (useMF ? 1 : 0) + (useVWAP ? 1 : 0);

  let signal, confidence;
  if (opposing > 0 && aligned === 0) {
    signal = 'oppose'; confidence = opposing >= 2 ? 'HIGH' : 'MEDIUM';
  } else if (aligned >= minComponents) {
    signal = 'agree';
    confidence = aligned >= maxComponents ? 'HIGH' : aligned >= minComponents ? 'MEDIUM' : 'LOW';
  } else {
    signal = 'neutral'; confidence = 'LOW';
  }

  return {
    signal,
    confidence,
    components:  aligned,
    opposing,
    wt:   { value: Math.round(wt1Val * 100) / 100, signal: wtSig },
    mf:   useMF    ? { value: Math.round(mfVal * 10) / 10, signal: mfSig }       : null,
    vwap: useVWAP  ? { signal: vwapSig, divergence: vwapDiv }                    : null,
    reason:   notes.join(' · ') || 'No alignment',
    warnings,
  };
}
